(EZ-70) Modify service 'start' functionality to use new TK 'start' subcommand

janelu2 opened the pull request:
Previously, system init scripts would daemonize a java process and then use a post-start command to poll for the port to open with netstat. This PR includes the new start script which will both daemonize a java process and poll for the service status using a restart-file implemented in [TK-345](https://github.com/puppetlabs/trapperkeeper/pull/248). The service scripts have been changed to call on this new `start` subcommand.

janelu2 commented:
also doing some git squashing/conflict fixing, please ignore the crazy # of commits

haus commented on resources/puppetlabs/lein-ezbake/template/foss/ext/debian/ezbake.init.erb:
```diff
@@ -58,37 +58,8 @@ EXEC="$JAVA_BIN -XX:OnOutOfMemoryError=\"kill -9 %p\" -Djava.security.egd=/dev/u
 do_start()
 {
-    <% if EZBake::Config[:open_file_limit] -%>
-    [ -n "$OPEN_FILE_LIMIT" ] && ulimit -n $OPEN_FILE_LIMIT
```
I think you may have deleted too much here.

janelu2 replied:
could I do this in the start script instead?

haus replied:
i'm not sure. I think it's easier to just leave here.

mckern replied:
I don't know if the code to check open files values should be left in or not; I don't see other init scripts from other vendors doing that, but nothing I checked starts java services.

haus replied:
the open file limit stuff was added recently in https://github.com/puppetlabs/ezbake/pull/321

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/debian/ezbake.init.erb:
```diff
@@ -58,37 +58,8 @@ EXEC="$JAVA_BIN -XX:OnOutOfMemoryError=\"kill -9 %p\" -Djava.security.egd=/dev/u
-
-    <% EZBake::Config[:debian][:pre_start_action].each do |action| -%>
-    <%= action %>
-    <% end -%>
```
Do we lose anything by removing the `pre_start_action` hook?

janelu2 replied:
This is the pre-start action: `:pre_start_action => ["/usr/bin/install --directory --owner=puppet --group=puppet --mode=775 /var/run/puppetlabs/puppetserver"]`

janelu2 replied:
@mckern do you think I should keep it? It looks like it is creating an install dir

janelu2 replied:
just an update, gonna keep this >_> I think it's kinda important

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/debian/ezbake.init.erb:
```diff
@@ -58,37 +58,8 @@ EXEC="$JAVA_BIN -XX:OnOutOfMemoryError=\"kill -9 %p\" -Djava.security.egd=/dev/u
-    if [ ! -d  "/var/run/puppetlabs/$NAME" ] ; then
-      mkdir -p /var/run/puppetlabs/$NAME
-      chown -R $USER:$GROUP /var/run/puppetlabs/$NAME
-    fi
```
Does TK now handle maintaining its own run directory/state?

janelu2 replied:
TK manages checking that the directory exists and creates it if it does not

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/debian/ezbake.init.erb:
```diff
@@ -58,37 +58,8 @@ EXEC="$JAVA_BIN -XX:OnOutOfMemoryError=\"kill -9 %p\" -Djava.security.egd=/dev/u
-    if [ "$retval" -eq 0 ]; then
-    <% EZBake::Config[:debian][:post_start_action].each do |action| -%>
-        <%= action %>
-    <% end -%>
```
See previous comment re: `pre_start_action` hooks, but applied to `post_start_action` instead.

haus replied:
pre and post start actions should stick around in general.

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/debian/ezbake.service.erb:
```diff
@@ -24,13 +25,7 @@ ExecStartPre=<%= action %>
-          -b "${BOOTSTRAP_CONFIG}" $@
+ExecStart=/opt/puppetlabs/server/apps/<%= EZBake::Config[:real_name] %>/bin/<%= EZBake::Config[:real_name] %> start
 
 KillMode=process
```
Is `KillMode` still `process` or should we now define an `ExecStop` value?

haus replied:
process is still good according to my read of the docs.

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/ezbake.service.erb:
```diff
@@ -22,19 +23,12 @@ PermissionsStartOnly=true
 ExecStartPre=<%= action %>
 <% end -%>
 <% end -%>
+
```
Not a deal breaker, but extra whitespace

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/ezbake.service.erb:
```diff
@@ -22,19 +23,12 @@ PermissionsStartOnly=true
 
 KillMode=process
 
-ExecStartPost=/bin/bash "${INSTALL_DIR}/ezbake-functions.sh" wait_for_app
```
See previous comment about post-start actions; also, I see that the `debian/ezbake.service.erb` file doesn't have any post-start hooks? Are we worried about consistency?

janelu2 replied:
I believe @haus knows more than me about this one.

haus replied:
This post should go away, as the start script removes the need for it. The generic templated list immediately below should stay.

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -59,44 +59,8 @@ find_my_pid() {
-
-    <% EZBake::Config[:redhat][:pre_start_action].each do |action| -%>
-    <%= action %>
-    <% end -%>
```
Previous comments re: `pre_start_action` and open file limits apply.

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -59,44 +59,8 @@ find_my_pid() {
-    # call status here and figure out current state
-    rh_status_q
-    [ -x $JAVA_BIN ] || exit 5
-    [ -e $config ] || exit 6
```
We can probably stand to lose the `JAVA_BIN` checking but do we want to lose the `$config` test? Does TK handle that now?

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -59,44 +59,8 @@ find_my_pid() {
-    [ -x $JAVA_BIN ] || exit 5
-    [ -e $config ] || exit 6
-    # Move any heap dumps aside
-    echo -n $"Starting $prog: "
```
No, keep this (or is TK verbose now?)

mckern commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -59,44 +59,8 @@ find_my_pid() {
-    if [ "$retval" -eq 0 ]; then
-    <% EZBake::Config[:redhat][:post_start_action].each do |action| -%>
-        <%= action %>
-    <% end -%>
```
Previous comments re: `post_start_action` apply

janelu2 commented:
@mckern I didn't realize I removed so much, I think I'm going to add back the pre, post, and open file limit check. I am not sure about the post in the service files though, I was told to remove those.

mckern commented on resources/puppetlabs/lein-ezbake/template/global/ext/cli/start.erb:
```diff
@@ -4,15 +4,25 @@ set -e
     mkdir -p "$dir"
     echo "1" > "$restartfile"
+else
+    if [ ! -r "$restartfile" ] || [ ! -w "$restartfile" ]; then
```
Can lines 21 and 22 be combined into `elif [ ! -r "$restartfile" ] || [ ! -w "$restartfile" ]; then`?

janelu2 replied:
@mckern yes I'll change that

mckern commented:
@janelu2 some of my comments apply to numerous files across the whole PR, and my insight into EZBake development is 💩 so please take comments or suggestions with a grain of salt. Let me know if/when you'd like a re-review and I'll do my best to be helpful.

haus commented on resources/puppetlabs/lein-ezbake/template/global/ext/cli/start.erb:
```diff
@@ -26,9 +36,15 @@ while [ "$cur" == "$initial" ] ;do
+find_my_pid
+
+pid=`cat $PIDFILE`
+source "${INSTALL_DIR}/ezbake-functions.sh"
```
This line and the next line should be removed. They accomplished exactly what the start replacement is supposed to be doing.

mckern replied:
What @haus said ^^

janelu2 commented:
@mckern @haus ok I brought back from the dead the lines that I mindlessly deleted. I gave it all a good read and RTFM-ing and I think I have it in much better shape now. Could you guys gimme another review :)?

mckern commented:
@janelu2 1 comment, but I like this. Fix it up and I am 👍 AF.

janelu2 commented:
Fixed the issue where the start script would time out and exit in waiting for the restart file to be incremented. This was done by changing the original script to be echoed a "0" with no newline. Previously, when "1" was echoed in, the service would start and write a "1" to the restart file as well, resulting in the start script to timeout/exit the service. 

Also I've tested this and it works :+1: on cent6, cent7, trusty, xenial, and precise

@haus @mckern @cprice404 
any other concerns/comments?

haus commented:
yes

haus commented:
@janelu2 in the refactoring dropping privileges to the specified user has been lost for systemv platforms (systemd looks okay). we'll need to add this back in.

janelu2 commented:
@haus thanks, forgot about that

janelu2 commented:
@haus I have updated all the scripts and tested them on sysv systems since they are the only ones affected by the latest commit. Will let you know if the suse one works out.. Any changes needed so far?

haus commented on resources/puppetlabs/lein-ezbake/template/foss/ext/debian/ezbake.init.erb:
```diff
@@ -66,7 +66,8 @@ do_start()
     <% end -%>
 
-    /opt/puppetlabs/server/apps/${REALNAME}/bin/${REALNAME} start >> /var/log/puppetlabs/${REALNAME}/${REALNAME}-daemon.log 2>&1
+    start-stop-daemon --start $EXTRA_ARGS --quiet --oknodo --pidfile $PIDFILE --chdir $INSTALL_DIR --exec $JAVA_BIN \
```
`$EXTRA_ARGS` includes `--background` and `--make-pidfile`, which seem unneeded/confusing with the start changes.

haus commented:
looking good aside from my recent comment. I'll do another pass tomorrow.

janelu2 commented:
@haus I've changed the args for `start-stop-daemon`, all the PID files to end in .pid, and added the logging for reload to display if it started correctly (like stop/start do).

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.erb:
```diff
@@ -135,7 +124,10 @@ reload() {
+
+    [ $RETVAL -eq 0 ] && success $"$base stopped" || failure $"$base reloaded"
     echo
+    [ $RETVAL -eq 0 ] && rm -f $lockfile $PIDFILE
```
You shouldn't delete the pidfile here.

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.erb:
```diff
@@ -135,7 +124,10 @@ reload() {
     /opt/puppetlabs/server/apps/${realname}/bin/${realname} reload
     RETVAL=$?
+
+    [ $RETVAL -eq 0 ] && success $"$base stopped" || failure $"$base reloaded"
```
missed a stopped => reloaded here

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.suse.erb:
```diff
@@ -170,6 +147,17 @@ reload() {
+    retval=$?
+
+    if [ "$retval" = 0 ]; then
+        log_success_msg $"${prog} stopped"
```
stopped => reloaded

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.suse.erb:
```diff
@@ -170,6 +147,17 @@ reload() {
+        echo
+        return 0
+    else
+        log_failure_msg $"${prog} could not be stopped, check ${LOGFILE}"
```
stopped => reloaded

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/debian/ezbake.init.erb:
```diff
@@ -66,21 +62,10 @@ do_start()
-
-    start-stop-daemon --start $EXTRA_ARGS --quiet --oknodo --pidfile $PIDFILE --chdir $INSTALL_DIR --exec $JAVA_BIN \
-      --startas /bin/bash -- -c "exec $EXEC >> /var/log/puppetlabs/${REALNAME}/${REALNAME}-daemon.log 2>&1"
+    start-stop-daemon --start --chuid $USER --oknodo --pidfile $PIDFILE --chdir $INSTALL_DIR \
```
would be good to put `--quiet` back here

haus commented on resources/puppetlabs/lein-ezbake/template/global/ext/cli/start.erb:
```diff
@@ -26,9 +35,10 @@ while [ "$cur" == "$initial" ] ;do
     ((timeout--))
     if [ "$timeout" = 0 ]; then
-        echo "Startup timed out after <%= EZBake::Config[:start_timeout] %> seconds"
+        echo "Startup timed out after <%= EZBake::Config[:start_timeout] %> seconds" >2
```
This will create a file called 2. should be a redirection to stderr instead `1>&2` iirc.

haus commented on resources/puppetlabs/lein-ezbake/template/global/ext/cli/start.erb:
```diff
@@ -4,15 +4,24 @@ set -e
-    echo "1" > "$restartfile"
+    echo -n "0" > "$restartfile"
+elif [ ! -r "$restartfile" ] || [ ! -w "$restartfile" ]; then
+        echo "The restart-file at <%= EZBake::Config[:restart_file] %> is not readable and/or writeable." >2
```
Same as below, use `1>&2` instead so you don't try to create a file called 2.

haus commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -135,7 +123,10 @@ reload() {
     /opt/puppetlabs/server/apps/${realname}/bin/${realname} reload
     RETVAL=$?
+
+    [ $RETVAL -eq 0 ] && success $"$base stopped" || failure $"$base reloaded"
```
stopped => reloaded

haus commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -135,7 +123,10 @@ reload() {
+
+    [ $RETVAL -eq 0 ] && success $"$base stopped" || failure $"$base reloaded"
     echo
+    [ $RETVAL -eq 0 ] && rm -f $lockfile $PIDFILE
```
we shouldn't delete pidfile here.

haus commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -74,21 +71,12 @@ start() {
     <%= action %>
     <% end -%>
 
-    pushd "${INSTALL_DIR}" &> /dev/null
```
The pushd might be important here. I forget why it was added initially.

haus replied:
I think if we try to start in a directory we don't own, bad things sometimes happen. Probably worth putting this back in.

haus replied:
also will need the matching popd below.

haus commented on resources/puppetlabs/lein-ezbake/template/foss/ext/debian/ezbake.init.erb:
```diff
@@ -66,21 +62,10 @@ do_start()
-
-    start-stop-daemon --start $EXTRA_ARGS --quiet --oknodo --pidfile $PIDFILE --chdir $INSTALL_DIR --exec $JAVA_BIN \
-      --startas /bin/bash -- -c "exec $EXEC >> /var/log/puppetlabs/${REALNAME}/${REALNAME}-daemon.log 2>&1"
+    start-stop-daemon --start --chuid $USER --oknodo --pidfile $PIDFILE --chdir $INSTALL_DIR \
```
would be good to put quiet back here

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.erb:
```diff
@@ -74,21 +71,13 @@ start() {
     <%= action %>
     <% end -%>
 
-    pushd "${INSTALL_DIR}" &> /dev/null
```
same as in foss, would be good to keep this pushd (and the matching popd).

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.suse.erb:
```diff
@@ -75,22 +74,10 @@ start() {
-    touch "${LOGFILE}"
-    chown $USER:$USER "${LOGFILE}"
-
-    export HOME="$(getent passwd ${USER} | cut -d':' -f6)"
```
We probably want to keep this around.

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.erb:
```diff
@@ -75,20 +72,13 @@ start() {
+    daemon --user $USER --pidfile $PIDFILE "/opt/puppetlabs/server/apps/<%= EZBake::Config[:real_name] %>/bin/<%= EZBake::Config[:real_name] %> start"
     popd &> /dev/null
-    find_my_pid
+    retval=$?
```
This should be above the popd, otherwise it will capture the return of the popd call.

haus commented on resources/puppetlabs/lein-ezbake/template/global/ext/cli/start.erb:
```diff
@@ -4,15 +4,24 @@ set -e
-    echo "1" > "$restartfile"
+    echo -n "0" > "$restartfile"
+elif [ ! -r "$restartfile" ] || [ ! -w "$restartfile" ]; then
+        echo "The restart-file at <%= EZBake::Config[:restart_file] %> is not readable and/or writeable." 1>&2
```
tabbing is off here

haus commented on resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/init.erb:
```diff
@@ -75,20 +72,13 @@ start() {
+    daemon --user $USER --pidfile $PIDFILE "/opt/puppetlabs/server/apps/${realname}/bin/${realname} start >> /var/log/puppetlabs/${realname}/${realname}-daemon.log 2>&1"
     popd &> /dev/null
-    find_my_pid
+    retval=$?
```
This should be above the popd, otherwise it will capture the return of the popd call.

haus commented:
@janelu2 i'm down to 4 comments

janelu2 commented:
@haus :sweat_smile:

haus commented on resources/puppetlabs/lein-ezbake/template/pe/ext/redhat/init.suse.erb:
```diff
@@ -81,16 +80,13 @@ start() {
+
     # startproc will change users, so make sure that user has permission
     # to access the present working directory.
-    cd "${INSTALL_DIR}"
```
this cd is important. it has the same purpose as the pushd/popd

haus commented:
I'm good here.

haus commented:
I'm just going to jot down a couple of notes here for myself. This shouldn't be part of the PR, but is stuff that came to mind while looking it over that we should look at implementing someday. I may ticket it.

- Consider passing -w to startproc, which should make it wait for the initial start script to return. not sure what it will do right now.
- Consider removing the -l and LOGFILE business from the sles init script. Logback should be doing the right thing there.
- Stop sourcing /lib/init/vars.sh in debian init scripts, or set VERBOSE to yes so that the service scripts will generate some output (they are currently totally silent).
- Consolidate usage of real_name, REALNAME, prog to be one thing. It is a giant PITA for anyone new to the codebase and makes reviewing changes harder.

haus commented:
@mckern would you or someone from @puppetlabs/release-engineering  like to take a pass at this?

underscorgan commented:
@janelu2 have you tested this? did everything seem like it was working?

if so, 👍 🎉

cprice404 commented:
@underscorgan FYI, @janelu2 's last day was yesterday so we'll have to pick up her work and push it forward.

She's done a fair amount of testing on this but I think it will be hard to test thoroughly until we can run through full PE test suites, etc.

This is what I think would probably be the best way forward:

* Go ahead and merge this
* Try to do some manual testing of consumption of this in OSS puppetserver via local beaker runs
* Assuming that goes well, go ahead and do a release here.  I kind of expect that there are going to be a few more bugs we won't have found prior to that but I think it will be a lot easier to find them once we have a versioned release.
* Integrate versioned release into OSS puppetserver, get tests passing, fix bugs if necessary.
* Try to do some manual testing of consumption of this in PE via local beaker runs, fix bugs if necessary.
* Integrate versioned release into PE puppetserver.

If that sounds okay to you, then I can merge this and try to keep moving things forward (or feel free to merge it when you're reading this if the plan sounds sane).

underscorgan commented:
sounds good to me, thanks @cprice404
